### ISSUE TITLE
Declare (global) variables that must be set before execution?

### ISSUE THREAD

**Issue by indigoviolet** [casey/just#1315]
Is it possible to declare variables that don't have default values assigned? 

The expected behavior would be that all recipes (that use the variable?) will fail to run without setting them at the command line. 

This would replicate the behavior with recipe parameters, but for global variables.

**Comment by casey** [comment]
That's a super interesting idea! It isn't currently implemented, but it seems like a reasonable feature, so I would accept a PR if anyone wants to implement it.

**Comment by indigoviolet** [comment]
Actually, an interesting approach is to use `env_var` and `env_var_or_default` liberally, which seems to address this issue as well as named parameters: #605 #1262 

```
foo := env_var("FOO")

bar quux=env_var("quux") baz=env_var_or_default("baz", "hi"):
    @echo {{ foo }} {{ quux }} {{ baz }}
```
```shell
❯ just -f test.justfile
error: Call to function `env_var` failed: environment variable `FOO` not present
  |
1 | foo := env_var("FOO")
  |        ^^^^^^^

❯ FOO=2 just -f test.justfile          
error: Call to function `env_var` failed: environment variable `quux` not present
  |
3 | bar quux=env_var("quux") baz=env_var_or_default("baz", "hi"):
  |          ^^^^^^^

❯ FOO=2 quux=3 just -f test.justfile
2 3 hi
```

In combination with `dotenv-load` and `--dotenv-filename`, it can make for very capable scripts. I wonder if a simpler path is to 

1. provide a simpler syntax for `env_var()` and `env_var_or_default`: I know we already use `$param` to indicate that it will be exported, but perhaps we could overload that to also mean it will be read from the environment if present.

Then we could write

```
bar $quux $baz="hi":
    @echo {{ foo }} {{ quux }} {{ baz }}
```

(I can't think of a clever syntax for the global variable equivalent that also supports default values)

2. Perhaps we can support `just -f test.justfile bar -- FOO=2 quux=3` to mean that those variables will be present in `just`'s environment.

**Comment by casey** [comment]
Good observation! The dollar sign in parameters can't be changed, because that would break existing justfiles. I do have #1278 open, which proposes shorter syntax for looking up environment variables. Either $FOO or `var("FOO")`. Maybe doing that is enough, and there doesn't need a special feature for this.

**Comment by indigoviolet** [comment]
>The dollar sign in parameters can't be changed, because that would break existing justfiles

> Parameters prefixed with a $ will be exported as environment variables

I'm not seeing why existing justfiles would break - `$foo` is currently a parameter that needs to be supplied/defaulted and it will be present in the environment of the recipe. The addition proposed here is that it can be supplied by setting it in the environment of the recipe. Why is that bad?

**Comment by casey** [comment]
> Why is that bad?

Because someone currently using `$foo` does not intend for it to fall back on the environment variable `foo` if it isn't found, they expect it to be an error if it isn't supplied.

Just must be extremely conservative in the changes it makes, since there are a lot of [justfiles in the wild](https://github.com/search?o=desc&q=filename%3Ajustfile&s=indexed&type=Code).

**Comment by indigoviolet** [comment]
Got it. Would it help if it were configurable and opt-in to start?

**Comment by casey** [comment]
I don't think so, it seems better that those two pieces of functionality remain separate: Exporting a variable to a recipe, and reading a variable from the environment. You want to be able to do neither, either, or both.
